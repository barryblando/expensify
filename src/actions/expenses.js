// import uuid from 'uuid';
import database from '../firebase/firebase';

// ------------------------------------------
// ASYNC ACTIONS FOR EXPENSE ARRAY REDUCER
// ------------------------------------------

// -- ADD_EXPENSE --
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense,
});

// By using redux thunk, you can be able to create asynchronous actions
// this is gonna get called by redux & can access dispatch and to dispatch addExpense
export const startAddExpense = (expenseData = {}) => (dispatch, getState) => {
  const uid = getState().auth.uid; // eslint-disable-line
  const { description = '', note = '', amount = 0, createdAt = 0 } = expenseData;
  const expense = { description, note, amount, createdAt };
  // push to firebase database
  // return promise for next .then success case ie. in test suite
  // otherwise test suite will show TypeError: Cannot read property 'then' of undefined
  return database
    .ref(`users/${uid}/expenses`)
    .push(expense)
    .then(ref => {
      // access ref to get the last key 'expenses/id' id = last part of the Reference's path & generated by firebase
      // dispatch addExpense to redux store for changes
      // don't forget that dispatch returns object
      dispatch(
        addExpense({
          id: ref.key,
          ...expense,
        })
      );
    });
};

// -- REMOVE_EXPENSE 'no default id required' --
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id,
});

export const startRemoveExpense = ({ id } = {}) => (dispatch, getState) => {
  const uid = getState().auth.uid; // eslint-disable-line
  return database
    .ref(`users/${uid}/expenses/${id}`)
    .remove()
    .then(() => dispatch(removeExpense({ id })))
    .catch(e => console.log(e));
};

// -- EDIT_EXPENSE --
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates,
});

export const startEditExpense = (id, updates) => (dispatch, getState) => {
  // get the current id logged in
  const uid = getState().auth.uid; // eslint-disable-line
  return database
    .ref(`users/${uid}/expenses/${id}`)
    .update(updates)
    .then(() => dispatch(editExpense(id, updates)));
};

// -- SET_EXPENSES --
export const setExpenses = expenses => ({
  type: 'SET_EXPENSES',
  expenses,
});

export const startSetExpenses = () => (dispatch, getState) => {
  // fetch all expense data once from firebase
  // return promise for next .then success case ie. in test suite
  // otherwise test suite will show TypeError: Cannot read property 'then' of undefined
  const uid = getState().auth.uid; // eslint-disable-line
  return database
    .ref(`users/${uid}/expenses`)
    .once('value')
    .then(snapshot => {
      // Parse the data into an array
      // const expenses = [];
      // </reference https://firebase.google.com/docs/reference/js/firebase.database.DataSnapshot#forEach
      // enumerating, creating & pushing new object into expenses array
      // snapshot.forEach(childSnapshot => {
      //   expenses.push({
      //     id: childSnapshot.key, // store the generated id by firebase to new id
      //     ...childSnapshot.val(), // spread out remaining value
      //   });
      // });
      const rawData = snapshot.val() || {};
      const expenses = Object.keys(rawData).reduce((prev, current) => {
        prev.push({ ...rawData[current], id: current });
        return prev;
      }, []);

      // Dispatch data to store for changes
      dispatch(setExpenses(expenses));
    })
    .catch(e => console.log(e));
};
