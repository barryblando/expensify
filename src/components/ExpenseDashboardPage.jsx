import React from 'react';
import ExpenseList from './ExpenseList'; /* eslint-disable-line */
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;
