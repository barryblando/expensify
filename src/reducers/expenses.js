import { ADD_EXPENSE, EDIT_EXPENSE, REMOVE_EXPENSE, SET_EXPENSES } from '../constants';

// ------------------------------------------
// Expenses Reducer
// ------------------------------------------
const expenseReducerDefaultState = [];

export default (state = expenseReducerDefaultState, action) => {
  switch (action.type) {
    case ADD_EXPENSE:
      return [...state, action.expense]; // using spread op same result as concat creating new array state rather than changing the prevState
    case REMOVE_EXPENSE:
      // state.map(expense => console.log(expense)); destruct expense array object only id in filter
      return state.filter(({ id }) => id !== action.id); // filtered out if not match
    case EDIT_EXPENSE:
      return state.map(expense => {
        // check if current expense iterating over by map match to action id
        if (expense.id === action.id) {
          // return brand new object & by using object spread operator
          return {
            ...expense, // grab all existing properties
            ...action.updates, // override any of the ones in expense that are pass down by updates
          };
        }
        return expense; // if no changes return
      });
    case SET_EXPENSES:
      // console.log(state);
      return action.expenses;
    default:
      return state;
  }
};
