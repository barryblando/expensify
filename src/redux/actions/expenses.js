import { push } from 'connected-react-router';
import database from '../../firebase/firebase';
import { ADD_EXPENSE, EDIT_EXPENSE, REMOVE_EXPENSE, SET_EXPENSES } from '../constants';

// --------------------------------------------------
// ASYNC ACTION CREATORS FOR EXPENSE ARRAY REDUCER
// --------------------------------------------------

// REDUX RULE INFO: Eliminate timing dependency bugs (No race between Async Request & View Render)

// -- ADD_EXPENSE --
export const addExpense = expense => ({
  type: ADD_EXPENSE,
  expense,
});

// In redux-thunk, we can now be able to create asynchronous action by using (HOF)
// whenever this thunk function gets triggered it's going to return a function and
// can access now to dispatch & getState and to dispatch normal(addExpense) action
export const startAddExpense = (expenseData = {}) => (dispatch, getState) => {
  const uid = getState().auth.uid; // eslint-disable-line
  const { description = '', note = '', amount = 0, createdAt = 0 } = expenseData;
  const expense = { description, note, amount, createdAt };
  // push to firebase database
  // return promise(database) for next .then success case ie. in test suite
  // otherwise test suite will show TypeError: Cannot read property 'then' of undefined
  return database
    .ref(`users/${uid}/expenses`)
    .push(expense)
    .then(ref => {
      // access ref to get the last key 'expenses/id' id = last part of the Reference's path & generated by firebase
      // dispatch addExpense to redux store for state updates and re-render views
      // don't forget that dispatch returns addExpense action object (this is for test suite)
      dispatch(
        addExpense({
          id: ref.key,
          ...expense,
        })
      );
      dispatch(push('/expense-dashboard'));
    });
};

// -- REMOVE_EXPENSE 'no default id required' --
export const removeExpense = ({ id } = {}) => ({
  type: REMOVE_EXPENSE,
  id,
});

export const startRemoveExpense = ({ id } = {}) => (dispatch, getState) => {
  const uid = getState().auth.uid; // eslint-disable-line
  return database
    .ref(`users/${uid}/expenses/${id}`)
    .remove()
    .then(() => {
      dispatch(removeExpense({ id }));
      dispatch(push('/expense-dashboard'));
    })
    .catch(e => console.log(e));
};

// -- EDIT_EXPENSE --
export const editExpense = (id, updates) => ({
  type: EDIT_EXPENSE,
  id,
  updates,
});

export const startEditExpense = (id, updates) => (dispatch, getState) => {
  // get the current id logged in
  const uid = getState().auth.uid; // eslint-disable-line
  return database
    .ref(`users/${uid}/expenses/${id}`)
    .update(updates)
    .then(() => {
      dispatch(editExpense(id, updates));
      // go back to dashboard
      dispatch(push('/expense-dashboard'));
    });
};

// -- SET_EXPENSES --
export const setExpenses = expenses => ({
  type: SET_EXPENSES,
  expenses,
});

export const startSetExpenses = () => (dispatch, getState) => {
  // fetch all expense data once from firebase
  const uid = getState().auth.uid; // eslint-disable-line
  return database
    .ref(`users/${uid}/expenses`)
    .once('value')
    .then(snapshot => {
      // Parse the data into an array
      // const expenses = [];
      // </reference https://firebase.google.com/docs/reference/js/firebase.database.DataSnapshot#forEach
      // enumerating, creating & pushing new object into expenses array
      // snapshot.forEach(childSnapshot => {
      //   expenses.push({
      //     id: childSnapshot.key, // store the generated id by firebase to new id
      //     ...childSnapshot.val(), // spread out remaining value
      //   });
      // });
      const rawData = snapshot.val() || {};
      // console.log('RawData: ', rawData);
      // </reference https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys
      const expenses = Object.keys(rawData).reduce((prev, current) => {
        prev.push({ ...rawData[current], id: current });
        // console.log('Previous: ', prev, 'Current: ', current);
        return prev;
      }, []);

      // Dispatch data to store for state updates and re-render views
      dispatch(setExpenses(expenses));
    })
    .catch(e => console.log(e)); // eslint-disable-line
};
