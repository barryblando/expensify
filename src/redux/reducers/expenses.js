import { ADD_EXPENSE, EDIT_EXPENSE, REMOVE_EXPENSE, SET_EXPENSES } from '../constants';

// ------------------------------------------
// Expenses Reducer (PURE FUNCTION)
// Make state transactional as possible
// ------------------------------------------
const expenseReducerDefaultState = [];

export default (state = expenseReducerDefaultState, action) => {
  switch (action.type) {
    case ADD_EXPENSE:
      // Make object & array immutable as possible
      // using spread operator to create new array state rather than mutating the prevState
      // & to prevent side effects. [...copying prevState, adding newObject]
      return [...state, action.expense];
    case REMOVE_EXPENSE:
      // create new expense array object using filter, return/filtered out if not match, otherwise don't include
      return state.filter(({ id }) => id !== action.id);
    case EDIT_EXPENSE:
      // return new constructed state array containing previous objects and the updated one
      return state.map(expense => {
        // check if current expense iterating over match to action id
        if (expense.id === action.id) {
          // return brand new object
          return {
            ...expense, // grab all properties from current expense object
            ...action.updates, // override the ones in current expense object that are pass down by updates
          };
        }
        return expense; // if no changes return
      });
    case SET_EXPENSES:
      return action.expenses;
    default:
      return state;
  }
};
