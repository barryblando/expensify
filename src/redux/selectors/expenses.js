import moment from 'moment';

// ------------------------------------------
// GET VISIBLE EXPENSES
// ------------------------------------------

// Use Selectors for Calculated State and Decoupling
// create a selector that simply exports all the variables needed to construct the view (e.g Filtering List)

export default (expenses, { text, sortBy, startDate, endDate }) =>
  // Filtering out
  expenses
    .filter(expense => {
      // Version 1.0
      // -------- LOGIC START --------
      // if startDate or endDate is a number then proceed to next condition (otherwise won't be filtered)
      // if createdAt is greater than startDate or less than endDate, do included in filter expenses (otherwise filtered out)
      // and if expenses.description has the text variable string inside of it (lowercase sensitive)
      // -------- LOGIC END --------
      // const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
      // const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
      // Version 2.0
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
      const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
      const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      let sortedValue;
      if (sortBy === 'date') {
        // <reference https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort
        sortedValue = a.createdAt < b.createdAt ? 1 : -1; // if true return b/1 else a/-1
      } else if (sortBy === 'amount') {
        sortedValue = a.amount < b.amount ? 1 : -1;
      }
      return sortedValue;
    });
