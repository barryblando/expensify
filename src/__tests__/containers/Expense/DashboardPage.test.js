import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import { expenses } from '../../fixtures/expenses';
import ExpenseDashboardPage from '../../../containers/Expense/DashboardPage';

const mockStore = configureMockStore();
const store = mockStore({});

test('should render Dashboard correctly', () => {
  const wrapper = shallow(
    <Provider store={store}>
      <ExpenseDashboardPage expenseCount={2} expensesTotal={243} expenses={expenses[0]} />
    </Provider>
  );
  expect(wrapper).toMatchSnapshot();
});
