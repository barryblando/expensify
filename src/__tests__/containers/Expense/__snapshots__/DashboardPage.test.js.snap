// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Dashboard correctly 1`] = `
<Connect(ExpenseDashboardPage)
  expenseCount={2}
  expenses={
    Object {
      "amount": 195,
      "createdAt": 0,
      "description": "Gum",
      "id": "1",
      "note": "",
    }
  }
  expensesTotal={243}
/>
`;
